src/main/java/org/elasticsearch/index/translog/fs/FsTranslog.java

```diff
@@ -189,168 +248,198 @@ public long ramBytesUsed() {
+                        + lastCommittedTranslogId + "], got [" + translogId + "]");
+            }
+            boolean found = false;
+            if (current.translogId() == translogId) {
+                found = true;
+            } else {
+                if (translogId > current.translogId()) {
+                    throw new ElasticsearchIllegalArgumentException("committed translog id must be lower or equal to current id (current ["
+                            + current.translogId() + "], got [" + translogId + "]");
                 }
             }
-            try {
-                newFile = type.create(shardId, id, new InternalChannelReference(location.resolve(getPath(id)), StandardOpenOption.READ, StandardOpenOption.WRITE, StandardOpenOption.CREATE_NEW), bufferSize);
-            } catch (IOException e) {
-                throw new TranslogException(shardId, "failed to create new translog file", e);
+            if (found == false) {
+                // try to find it in uncommittedTranslogs
+                for (int i = uncommittedTranslogs.size() - 1; i >= 0; i--) {
+                    if (uncommittedTranslogs.get(i).translogId() == translogId) {
+                        found = true;
+                        break;
+                    }
+                }
             }
-            FsTranslogFile old = current;
-            current = newFile;
-            IOUtils.close(old);
-        } finally {
-            rwl.writeLock().unlock();
-        }
-    }
+            if (found == false) {
+                ArrayList<Long> currentIds = new ArrayList<>();
+                for (FsChannelReader old : uncommittedTranslogs) {
+                    currentIds.add(old.translogId());
+                }
+                currentIds.add(current.translogId());
+                throw new FileNotFoundException("committed translog id can not be found (current ["
+                        + Strings.collectionToCommaDelimitedString(currentIds) + "], got [" + translogId + "]");
 
-    @Override
-    public void newTransientTranslog(long id) throws TranslogException {
-        rwl.writeLock().lock();
-        try {
-            assert this.trans == null;
-            long size = Long.MAX_VALUE;
-            Path location = null;
-            for (Path file : locations) {
-                long currentFree = Files.getFileStore(file).getUsableSpace();
-                if (currentFree < size) {
-                    size = currentFree;
-                    location = file;
-                } else if (currentFree == size && ThreadLocalRandom.current().nextBoolean()) {
-                    location = file;
+            }
+            lastCommittedTranslogId = translogId;
+            if (uncommittedTranslogs.isEmpty()) {
+                return;
+            }
+            while (uncommittedTranslogs.isEmpty() == false && uncommittedTranslogs.get(0).translogId() < translogId) {
+                FsChannelReader old = uncommittedTranslogs.remove(0);
```

Comment:
instead of doing this you can just safe the index of the one you wanna commit and then call
`IOUtils.close(list.subList(0, to)).clear();` it will all be in-place and only copies once under the hood to compact the list?